code/game/machinery/dance_machine.dm

```diff
@@ -0,0 +1,438 @@
+			if(25)
+				S.pixel_y = 7
+				S.forceMove(get_turf(src))
+		sleep(7)
+	if(selection.song_name == "Engineering's Ultimate High-Energy Hustle")
+		sleep(280)
+	for(var/obj/reveal in sparkles)
+		reveal.alpha = 255
+	while(active)
+		for(var/obj/item/device/flashlight/spotlight/glow in spotlights) // The multiples reflects custom adjustments to each colors after dozens of tests
+			if(QDELETED(src) || !active || QDELETED(glow))
+				return
+			if(glow.light_color == "red")
+				glow.light_color = "nw"
+				glow.light_power = glow.light_power * 1.48
+				glow.light_range = 0
+				glow.update_light()
+				continue
+			if(glow.light_color == "nw")
+				glow.light_color = "green"
+				glow.light_range = glow.range * 1.1
+				glow.light_power = glow.light_power * 2 // Any changes to power must come in pairs to neutralize it for other colors
+				glow.update_light()
+				continue
+			if(glow.light_color == "green")
+				glow.light_color = "sw"
+				glow.light_power = glow.light_power * 0.5
+				glow.light_range = 0
+				glow.update_light()
+				continue
+			if(glow.light_color == "sw")
+				glow.light_color = "purple"
+				glow.light_power = glow.light_power * 2.27
+				glow.light_range = glow.range * 1.15
+				glow.update_light()
+				continue
+			if(glow.light_color == "purple")
+				glow.light_color = "se"
+				glow.light_power = glow.light_power * 0.44
+				glow.light_range = 0
+				glow.update_light()
+				continue
+			if(glow.light_color == "se")
+				glow.light_color = "#ffff00"
+				glow.light_range = glow.range * 0.9
+				glow.update_light()
+				continue
+			if(glow.light_color == "#ffff00")
+				glow.light_color = "ne"
+				glow.light_range = 0
+				glow.update_light()
+				continue
+			if(glow.light_color == "ne")
+				glow.light_color = "red"
+				glow.light_power = glow.light_power * 0.68
+				glow.light_range = glow.range * 0.85
+				glow.update_light()
+				continue
+		if(prob(2))
+			INVOKE_ASYNC(src, .proc/hierofunk)
+		sleep(selection.song_beat)
+
+
+/obj/machinery/disco/proc/dance(var/mob/living/carbon/M) //Show your moves
+	switch(rand(0,9))
+		if(0 to 1)
+			dance2(M)
+		if(2 to 3)
+			dance3(M)
+		if(4 to 6)
+			dance4(M)
+		if(7 to 9)
+			dance5(M)
+	animate(M, transform = null, time = 1, loop = 0)
+
+/obj/machinery/disco/proc/dance2(var/mob/living/carbon/M)
+	set waitfor = 0
+	for(var/i = 1, i < 8, i++)
+		M.SpinAnimation(7,1)
+		M.setDir(pick(cardinal))
+		sleep(10)
+
+/obj/machinery/disco/proc/dance3(var/mob/living/carbon/M)
+	set waitfor = 0
+	for(var/i in 1 to 6)
+		if (!M)
+			return
+		M.SpinAnimation(7,1)
+		M.setDir(pick(cardinal))
+		for (var/x in 1 to 12)
+			sleep(1)
+			if (!M)
+				return
+			if (i<5)
+				M.pixel_y += 1
+			if (i>4)
+				M.pixel_y -= 2
+			M.setDir(turn(M.dir, 90))
+			switch (M.dir)
+				if (NORTH)
+					M.pixel_y += 3
+				if (SOUTH)
+					M.pixel_y -= 3
+				if (EAST)
+					M.pixel_x -= 3
+				if (WEST)
+					M.pixel_x += 3
+		sleep(12)
+	M.pixel_x = 0
+	M.pixel_y = 0
```

Comment:
I mean, initial would work for the dragon but not for people lying down, you might have to cut this for other, safer/better controlled effects.